### ISSUE TITLE
My Precision and Recall, Map Value are always zero

### ISSUE THREAD

**Issue by MarooNarae** [ultralytics/yolov5#10739]
### Search before asking

- [x] I have searched the YOLOv5 [issues](https://github.com/ultralytics/yolov5/issues) and [discussions](https://github.com/ultralytics/yolov5/discussions) and found no similar questions.

### Question

Hi. I was training  numberplates, and i labeled 175 images.

I've tried 300 epochs, but the Precision and Recall and Map Value are zero.

Is this my requirements or version issue?

It seems like detection with COCO works well.

![화면 캡처 2023-01-11 061434](https://user-images.githubusercontent.com/118100766/211664032-85f68ce2-86cf-4041-beff-9b78a031e2c3.png)
![results](https://user-images.githubusercontent.com/118100766/211664151-6aa9a404-2454-4ab9-a02a-7c891301fa2b.png)

**Comment by UNeedCryDear** [comment]
add --single-cls flag when training

**Comment by Air000** [comment]
Can you provide your training command?

**Comment by UNeedCryDear** [comment]
> Can you provide your training command?

This flag  needs to be added when there is only one class.
The order is the same as before, but add --single-cls.
example:
befor: ```python segment/train.py --data coco128-seg.yaml --weights yolov5s-seg.pt --img 640```
new: ```python segment/train.py --data coco128-seg.yaml --weights yolov5s-seg.pt --img 640 --single-cls```

**Comment by Air000** [comment]
> > Can you provide your training command?
> 
> This flag needs to be added when there is only one class. The order is the same as before, but add --single-cls. example: before: `python segment/train.py --data coco128-seg.yaml --weights yolov5s-seg.pt --img 640` new: `python segment/train.py --data coco128-seg.yaml --weights yolov5s-seg.pt --img 640 --single-cls`

From the log, no classes were detected in the validating run, the training loss data are all zero. I guess the issue is with the dataset. Here are some suggestions for further checking:
1. Check on the full training log to make sure the 173 images were scanned. There should be a line `train: Scanning /content/datasets/coco128-seg/labels/train2017... 126 images, 2 backgrounds, 0 corrupt: 100% 128/128 [00:00<00:00, 1389.59it/s]` ; Check on your dataset and make sure the .yaml points to the right path. 
2. Read the [official segmentation tutorial](https://github.com/ultralytics/yolov5/blob/master/segment/tutorial.ipynb)

**Comment by MarooNarae** [comment]
my training command was

python train.py --data test.yaml --weights yolov5s.pt --batch 16 --epochs 300 --cfg yolov5s.yaml 

I will try with adding --single-cls

**Comment by MarooNarae** [comment]
> > > Can you provide your training command?
> > 
> > 
> > This flag needs to be added when there is only one class. The order is the same as before, but add --single-cls. example: before: `python segment/train.py --data coco128-seg.yaml --weights yolov5s-seg.pt --img 640` new: `python segment/train.py --data coco128-seg.yaml --weights yolov5s-seg.pt --img 640 --single-cls`
> 
> From the log, no classes were detected in the validating run, the training loss data are all zero. I guess the issue is with the dataset. Here are some suggestions for further checking:
> 
> 1. Check on the full training log to make sure the 173 images were scanned. There should be a line `train: Scanning /content/datasets/coco128-seg/labels/train2017... 126 images, 2 backgrounds, 0 corrupt: 100% 128/128 [00:00<00:00, 1389.59it/s]` ; Check on your dataset and make sure the .yaml points to the right path.
> 2. Read the [official segmentation tutorial](https://github.com/ultralytics/yolov5/blob/master/segment/tutorial.ipynb)

I checked full training log, and there is a warning.

WARNING  Cache directory D:\YOLO_V5\yolov5\test is not writeable: [WinError 183]      : 'D:\\YOLO_V5\\yolov5\\test\\labels.cache.npy' -> 'D:\\YOLO_V5\\yolov5\\test\\labels.cache'

So this may be a problem. I will check this too.

**Comment by UNeedCryDear** [comment]
delete D:\YOLO_V5\yolov5\test\labels.cache

**Comment by MarooNarae** [comment]
> delete D:\YOLO_V5\yolov5\test\labels.cache

I tried once again, but the values are still all zero.
I think labeling is not going right.
I made the txt files that contains path of image files, should I put that in yaml file?

**Comment by UNeedCryDear** [comment]
Your ```test.yaml``` file should be changed according to the format of ```data/coco.yaml```, and ensure that your image and annotation file are the same as the coco data set.
![image](https://user-images.githubusercontent.com/52729998/211757784-6d9a82cc-d7df-401f-ade3-0a4e4fe52d88.png)

**Comment by MarooNarae** [comment]
> Your `test.yaml` file should be changed according to the format of `data/coco.yaml`, and ensure that your image and annotation file are the same as the coco data set. ![image](https://user-images.githubusercontent.com/52729998/211757784-6d9a82cc-d7df-401f-ade3-0a4e4fe52d88.png)

Well, my bad. I can't understand well.

yolov5
- test
  - images
  - labels
  
I have test folder under the yolov5 folder, and images and labels folder inside test folder. 
Can you just check that my test.yaml is right?? I'm still confused now... :(
![화면 캡처 2023-01-11 061434](https://user-images.githubusercontent.com/118100766/211760316-a6bd8067-c26d-465f-a586-477ffe8c0aec.png)

**Comment by UNeedCryDear** [comment]
Can you show me a label file? If you are based on coco.yaml, you have to follow the format of coco. But it seems that your writing is based on coco128.yaml. I don't know what your specific label format is, and I don't know how to modify it. Or you need to modify your annotation according to this tutorial.
https://docs.ultralytics.com/yolov5/tutorials/train_custom_data#1-create-dataset

**Comment by MarooNarae** [comment]
![화면 캡처 2023-01-11 061434](https://user-images.githubusercontent.com/118100766/211772754-cb496a72-c228-49e6-8f07-4dba2087b408.png)
This is one of my label file, I labeled it with labelImg...!

**Comment by MarooNarae** [comment]
![labels](https://user-images.githubusercontent.com/118100766/211773912-2cec69bb-6e24-4a6f-a1de-55de0757db67.jpg)

+ Does the red box means that I failed in labeling??

**Comment by UNeedCryDear** [comment]
> This is one of my label file, I labeled it with labelImg...!

this is coco128 format.

> Does the red box means that I failed in labeling?

Your picture shows that your labels were read correctly.

If you add ```--single-cls```and it still doesn't work, I don't know what the problem is.

**Comment by MarooNarae** [comment]
> this is coco128 format.

Thank you for your support...! I appreciate your effort 👍

**Comment by Air000** [comment]
@MarooNarae Have you checked on this thread [_yolov5 does not train or detect_](https://github.com/ultralytics/yolov5/issues/4839#top). It seems similar to yours. The issue is about the version of cuDNN in PyTorch. 
The solution is in [this comment](https://github.com/ultralytics/yolov5/issues/4839#issuecomment-932041190), which is:
`pip3 install torch==1.9.1+cu102 torchvision==0.10.1+cu102 torchaudio===0.9.1 -f https://download.pytorch.org/whl/torch_stable.html`

Before copying and pasting the solution directly, please check your installation command according to your cuda version on [here](https://pytorch.org/).

Can you send the complete training log here if the problem still exists, so that someone can find any clues from the log.

**Comment by MarooNarae** [comment]
> @MarooNarae Have you checked on this thread [_yolov5 does not train or detect_](https://github.com/ultralytics/yolov5/issues/4839#top). It seems similar to yours. The issue is about the version of cuDNN in PyTorch. The solution is in [this comment](https://github.com/ultralytics/yolov5/issues/4839#issuecomment-932041190), which is: `pip3 install torch==1.9.1+cu102 torchvision==0.10.1+cu102 torchaudio===0.9.1 -f https://download.pytorch.org/whl/torch_stable.html`
> 
> Before copying and pasting the solution directly, please check your installation command according to your cuda version on [here](https://pytorch.org/).
> 
> Can you send the complete training log here if the problem still exists, so that someone can find any clues from the log.

thank you for your answer. 
I erased all of my environments and tried with cuda 10.2, and it works fine.
Really thank you for all your effort..!!

**Comment by Air000** [comment]
> thank you for your answer. I erased all of my environments and tried with cuda 10.2, and it works fine. Really thank you for all your effort..!!

Since the problem is solved, can you close this issue? Thanks.

**Comment by glenn-jocher** [comment]
Dear @Air000,

Glad to hear that the issue is resolved! If you encounter any other challenges or have additional questions, please don't hesitate to reach out. Your feedback and contributions are greatly appreciated! Thank you for your patience and understanding.

Best regards.
